routing/dht/dht.go

```diff
@@ -279,9 +289,24 @@ func (dht *IpfsDHT) findProvidersSingle(ctx context.Context, p peer.ID, key key.
 
 // nearestPeersToQuery returns the routing tables closest peers.
 func (dht *IpfsDHT) nearestPeersToQuery(pmes *pb.Message, count int) []peer.ID {
-	key := key.Key(pmes.GetKey())
-	closer := dht.routingTable.NearestPeers(kb.ConvertKey(key), count)
-	return closer
+	if pmes.Key != nil {
+		return dht.routingTable.NearestPeers(kb.ConvertKey(key.Key(pmes.GetKey())), count)
+	} else {
+		closer := make(map[peer.ID]struct{})
+		for _, sk := range pmes.GetKeys() {
+			np := dht.routingTable.NearestPeers(kb.ConvertKey(key.Key(sk)), count)
+			for _, p := range np {
+				closer[p] = struct{}{}
+			}
+		}
+
+		var out []peer.ID
+		for p, _ := range closer {
+			out = append(out, p)
+		}
+
+		return out
```

Comment:
- ok what you say makes sense. 
- since the returned list is not associated to each key (doesnt need to be, requestor can do it) this is starting to look like "give me your routing table", for large queries. (that -- btw -- maybe should be a method on the dht)
- perhaps rename this function? a "query" so far has been a process spanning multiple request/response pairs searching for a single endpoint. kademlia and most dhts uses "query" to represent request for a single endpoint too. i dont think we should reshape to mean that a single query could have multiple endpoints-- instead a single request/response message can involve multiple queries simultaneously. 
-  i suggest `nearestPeersToAllKeys` or something equally descriptive
- ok i still think what i said is true about _the query itself_ in general, is it not? i refer to now shotgunning around the dht, without sorting by nearest
- and i still think my two suggestions for coalescing above are very useful